HPCC-26590 Variable/delimitted spray fails if the first record/row is longer than the default block size (256k).

AttilaVamos opened the pull request:
Fix first record size issue in CSV quick partitioner.

Originally the parameter of the ensureBuffered() function was blockSize  and it assumed the first record is not too big and ignored the record size parameter. It was wrong.
The new parameter is the bufferSize what is calculated from the blockSize and expectedRecordSize

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->
Tested manually with the user provided file
<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

AttilaVamos committed:
HPCC-26590 Variable/delimitted spray falls if the first record/row is longer than the default block size (256k).

Fix first record size issue in CSV quick partitioner.

Tested manually with the user provided file

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

HPCCSmoketest commented:
Process of PR-15440, label: HPCC-26590-fix-8.2.x starts now.
The reason of this test is: Forced to re-test.
Commit ID: 6568d738d1536535f8b024cf6f4670b16ae88886
Estimated completion time is 0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-224.ca-central-1.compute.internal
 Git:  git version 2.9.5

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-224.ca-central-1.compute.internal
 Git:  git version 2.9.5
Sha: 6568d738d1536535f8b024cf6f4670b16ae88886
Containerized:False
Build: success
Milestone:Install hpccsystems-platform-community_8.2.25-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |169 |169 |0 |0 |0 |69 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |6 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |9 |9 |0 |25 sec  (00:00:25) |
|setup (thor) |9 |9 |0 |39 sec  (00:00:39) |
|setup (roxie) |19 |19 |0 |20 sec  (00:00:20) |
|test (hthor) |982 |982 |0 |916 sec  (00:15:16) |
|test (thor) |892 |892 |0 |1027 sec  (00:17:07) |
|test (roxie) |1051 |1051 |0 |967 sec  (00:16:07) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    8 sec (00:00:08) |  556 sec (00:09:16) |  126 sec (00:02:06) |   22 sec (00:00:22) |   16 sec (00:00:16) | 1354 sec (00:22:34) |   19 sec (00:00:19) | 2101 sec (00:35:01) |

jakesmith commented on dali/ft/daftformat.cpp:
```diff
@@ -1065,11 +1065,10 @@ void CCsvQuickPartitioner::findSplitPoint(offset_t splitOffset, PartitionCursor
-        if (format.maxRecordSize + maxElementLength > blockSize)
-            eof = !ensureBuffered(blockSize);
-        else
-            eof = !ensureBuffered(format.maxRecordSize + maxElementLength);
```
The above condition looks backward, but correct if flipped ?
i.e. if > blockSize should be using format.maxRecordSize + maxElementLength

jakesmith replied:
also, the same code is present in a few other places, if this is wrong, so are other places I think.

jakesmith commented on dali/ft/daftformat.cpp:
```diff
@@ -1065,11 +1065,10 @@ void CCsvQuickPartitioner::findSplitPoint(offset_t splitOffset, PartitionCursor
-            eof = !ensureBuffered(format.maxRecordSize + maxElementLength);
+        // Originally the parameter of the ensureBuffered() function was blockSize
+        // and it assumed the first record is not too big and ignored the record size parameter. It was wrong.
+        // The new parameter is the bufferSize what is calculated from the blockSize and expectedRecordSize
```
Shouldn't really add comments to code referrring to code that no longer exists,
but should be in PR/JIRA description/comments.

AttilaVamos replied:
Removed and added to PR description

jakesmith reviewed (commented):
@AttilaVamos - please see comments.

Also there's a typo in commit/PR message : falls->fails, and PR check boxes not checked.

AttilaVamos committed:
Fix the partions creation without separately handled the first part
of the file.

Fix the invalid (-1) wichOutput index generation for the first part.

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

jakesmith commented on dali/ft/daftformat.cpp:
```diff
@@ -143,8 +143,11 @@ void CPartitioner::commonCalcPartitions()
-            results.append(*new PartitionPoint(whichInput, split-1, startInputOffset-thisOffset+thisHeaderSize, inputOffset - startInputOffset - cursor.trimLength, cursor.outputOffset-startOutputOffset));
-            JSON_DBGLOG("commonCalcPartitions: startInputOffset:%lld, inputOffset: %lld, startOutputOffset: %lld, cursor.outputOffset: %lld",startInputOffset ,inputOffset, startOutputOffset, cursor.outputOffset);
+            // To avoid the first partition's whichOutput value becomes -1 (which means invalid)
+            unsigned whichOutput = (split > 0 ? split-1 : split);
```
when does it enter this block, such that split is 0?
I thought the if (split != firstSplit) covered that case?
Perhaps comment need extending to explain.

AttilaVamos replied:
After some tracing and debugging I find out when the split == 0 the partition out file index becomes “-1” what means invalid/not defined file. For usual files (with multiple records what can distribute among the target parts) the split moving from first split to the last one during the partition creation, but in a special occasion e.g. when there is only one record for a multiple target parts environment, the scan can reach the end of the record (and the input file) and it is still in the split ==0 state.

jakesmith commented on dali/ft/daftformat.cpp:
```diff
@@ -1002,79 +1005,64 @@ void CCsvQuickPartitioner::findSplitPoint(offset_t splitOffset, PartitionCursor
 {
     const byte *buffer = bufferBase();
     numInBuffer = bufferOffset = 0;
-    if (splitOffset != 0)
```
git not making it easy to see the change, but I see it's just removal of this if and the else block.
tbh, I'm not familiar with this code, and am not sure what it was doing (or what it thought it was doing) in the case of splitOffset == 0 vs not.
@ghalliday - could you also review this please?

AttilaVamos replied:
I gave my explanation below why I removeed this condition and the related else branch.

jakesmith commented on dali/ft/daftformat.cpp:
```diff
@@ -1339,81 +1327,67 @@ void CUtfQuickPartitioner::findSplitPoint(offset_t splitOffset, PartitionCursor
 {
     const byte *buffer = bufferBase();
     numInBuffer = bufferOffset = 0;
-    if (splitOffset != 0)
```
git not making it easy to see the change, but I see it's just removal of this if and the else block.
@ghalliday - would you look at this change too please?

jakesmith commented on dali/ft/filecopy.cpp:
```diff
@@ -830,8 +830,12 @@ void FileSprayer::assignPartitionFilenames()
-        cur.outputName.set(targets.item(cur.whichOutput).filename);
-        setCanAccessDirectly(cur.outputName);
+        // Prevent to use invalid (-1) whichOutput index.
+        if (cur.whichOutput != (unsigned)-1)
```
I am not entirely sure at the moment, how a PartitionPoint with a whichOutput of -1 is getting added at the moment,
hence I'm not sure if this change should be necessary. See my other comment in CPartitioner::commonCalcPartitions() in daftformt.cpp

AttilaVamos replied:
Changed to throw and error.

jakesmith reviewed (commented):
@AttilaVamos - a couple of questions.
But I'd also like @ghalliday to take a look.

ghalliday reviewed (commented):
@AttilaVamos I don't think the fix is right for a few reasons.

Previously are headers included in the sprayed file?  I think this change would also strip them.

AttilaVamos committed:
WIP

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

AttilaVamos committed:
WIP add regression tests.

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

HPCCSmoketest commented:
Process of PR-15440, label: HPCC-26590-fix-8.2.x starts now.
The reason of this test is: Forced to re-test.
Commit ID: 69109f39c836aebc02e3c21a1961ea5d3469cf9d
Estimated completion time is 0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-101.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.18.0,  cUrl: 7.67.0,  node.js: v16.13.0,  npm: 8.1.0

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-101.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.18.0,  cUrl: 7.67.0,  node.js: v16.13.0,  npm: 8.1.0
Sha: 69109f39c836aebc02e3c21a1961ea5d3469cf9d
Containerized:False
Build: success
Milestone:Install hpccsystems-platform-community_8.2.25-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |169 |169 |0 |0 |0 |60 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |4 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |9 |9 |0 |22 sec  (00:00:22) |
|setup (thor) |9 |9 |0 |38 sec  (00:00:38) |
|setup (roxie) |19 |19 |0 |18 sec  (00:00:18) |
|test (hthor) |982 |982 |0 |903 sec  (00:15:03) |
|test (thor) |930 |930 |0 |1147 sec  (00:19:07) |
|test (roxie) |1051 |1051 |0 |1093 sec  (00:18:13) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   15 sec (00:00:15) |  476 sec (00:07:56) |   97 sec (00:01:37) |   18 sec (00:00:18) |   16 sec (00:00:16) | 1457 sec (00:24:17) |   17 sec (00:00:17) | 2096 sec (00:34:56) |

HPCCSmoketest commented:
Process of PR-15440, label: HPCC-26590-fix-8.2.x starts now.
The reason of this test is: Forced to re-test.
Commit ID: 69109f39c836aebc02e3c21a1961ea5d3469cf9d
Estimated completion time is 0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-159.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.18.0,  cUrl: 7.67.0,  node.js: v16.13.0,  npm: 8.1.0

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-159.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.18.0,  cUrl: 7.67.0,  node.js: v16.13.0,  npm: 8.1.0
Sha: 69109f39c836aebc02e3c21a1961ea5d3469cf9d
Containerized:False
Build: success
Milestone:Install hpccsystems-platform-community_8.2.25-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |169 |169 |0 |0 |0 |60 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |5 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |9 |9 |0 |22 sec  (00:00:22) |
|setup (thor) |9 |9 |0 |39 sec  (00:00:39) |
|setup (roxie) |19 |19 |0 |18 sec  (00:00:18) |
|test (hthor) |982 |982 |0 |838 sec  (00:13:58) |
|test (thor) |907 |907 |0 |957 sec  (00:15:57) |
|test (roxie) |1051 |1051 |0 |847 sec  (00:14:07) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   16 sec (00:00:16) |  477 sec (00:07:57) |   98 sec (00:01:38) |   19 sec (00:00:19) |   16 sec (00:00:16) | 1269 sec (00:21:09) |   18 sec (00:00:18) | 1913 sec (00:31:53) |

AttilaVamos commented:
@jake @ghalliday 
It is hard to explain how findSplitPoint() function works and why the changes may be good based on  some grabbed out lines of code.
First of all the “if (splitOffset != 0)” condition and the related else block introduced when the initial version of CSV record structure discovery (create a naïve/basic ECL record structure with matching number of String/UTF8 fields)  or load the record structure defined in the first record of CSV file (if recordStructurePresent flag is true) implemented and the aim of that condition and else block was to process first record individually.
Later, when the processFullBuffer introduced to improve performance, to handling the record structure discovery/load is changed and it wasn’t necessary to handle the first record separately anymore. It is shame that “if – else” branch didn’t removed at that time. 
However this separated first record handling didn’t caused any problem until it faced a file with large (>256kB) first record. The problem was it reads the first 256kB (buffer size) of the file and it ignores the hint comes from the recordSize parameter.  So, the large first record caused exception and even though the user specified a larger record size value, it was not used.
My first fix was to modify that else branch to use recordSize parameter to read the first record, but Jake’s first review comments pointed out it wasn’t clear. Then I recheck the whole findSplitPoint() function and find out couple of things:
1. My first changes can cause large buffer allocation for the first record – to avoid it I need implement some loop to process the first record in chunks. But it is already implemented in the true branch.
2. Record structure handling doesn’t require separate first record handle. 
So I removed the “if(splitOffset != 0)” and the related else branch but this  change caused spray error related the output file. After some tracing and debugging I find out when the splitOffset == 0 the partition out file index becomes “-1” what means invalid/not defined file.  For usual files (with multiple records what can distribute among the target parts) the split offset moving from first split to the last one during the partition creation, but in a special occasion e.g. when there is only one large record for a multiple target parts environment, the scan can reach the end of the record (and the input file) and it is still in the splitOffset==0 state. 
A spray_test_csv.ecl regression test created to test several special occasions, like only one small, medium, large record in the input file, one small and one large/huge record in input with alternated sequence and so on.  To keep it quick (to enamel to run in Smoketest) it has been split into two test cases: spray_test_csv.ecl and spray_test_csv_stress.csv (marked with ‘//class=stress’).  The first executed in every environment (Smoketest, OBT), but the latter is only in OBT (based on the longer execution time)

ghalliday reviewed (commented):
@AttilaVamos A couple of comments, and a few comments from before that you haven't responded to.
I am a bit concerned that the way the record header code is implemented will lead to additional reads from files - with a significant cost overhead reading on azure.  I have opened a separate jira HPCC-26844 for that.

ghalliday commented on dali/ft/daftformat.cpp:
```diff
@@ -1002,79 +1005,64 @@ void CCsvQuickPartitioner::findSplitPoint(offset_t splitOffset, PartitionCursor
-        else if (splitOffset - thisOffset < thisSize)
-            throwError2(DFTERR_UnexpectedReadFailure, fullPath.get(), splitOffset-thisOffset+thisHeaderSize);
-    }
-    else
```
I think it is probably more subtle than that - you may lose the first characters of the header if multi character line terminators are used.  The fix suggest above should avoid the problem.

AttilaVamos replied:
Hmm... I am not 100% sure about what you mean. I need to recheck this.

jakesmith reviewed (commented):
@AttilaVamos - I have no additional comments over Gavin and my previous comments, please address/respond to those.

AttilaVamos commented on dali/ft/daftformat.cpp:
```diff
@@ -1002,79 +1006,64 @@ void CCsvQuickPartitioner::findSplitPoint(offset_t splitOffset, PartitionCursor
     {
-        seekInput(splitOffset-thisOffset+thisHeaderSize);
+        //Throw away the first match, incase we hit the \n of a \r\n or something similar.
+        if (maxElementLength > 1)
```
At the moment I can't see why 'if splitOffset != 0', but I will trace this part again to check.

ghalliday commented:
Timing this PR out.  Please reopen if comments are addressed
